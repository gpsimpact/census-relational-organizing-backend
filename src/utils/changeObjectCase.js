import { mapKeys, camelCase, snakeCase } from "lodash";

export const toCamelCase = obj => {
  if (!obj) return obj;
  return mapKeys(obj, (value, key) => {
    return camelCase(key);
  });
};

export const toSnakeCase = obj =>
  mapKeys(obj, (value, key) => {
    return snakeCase(key);
  });
