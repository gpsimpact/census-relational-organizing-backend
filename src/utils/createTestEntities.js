import faker from "faker";
import { createGDS } from "@jakelowen/sqorn-graphql-filters";
import { sq } from "../db";

export const uuid = () => faker.random.uuid();

export const createTestUser = async (
  name,
  email,
  abbreviation,
  active = true
) =>
  createGDS(sq.from`users`)({
    name: name || `${faker.name.firstName()} ${faker.name.lastName()}`,
    email: email || faker.internet.email().toLowerCase(),
    abbreviation: faker.lorem.word().substring(0, 7),
    active
  });

export const createTestGlobalPerm = async (userId, permission) =>
  createGDS(sq.from`global_permissions`)({
    userId,
    permission
  });

export const createTestClient = async (active = true) =>
  createGDS(sq.from`clients`)({
    name: faker.company.companyName(),
    abbreviation: faker.random.alphaNumeric(7),
    slug: faker.lorem.slug(),
    active
  });

export const createTestOLPermission = async (userId, clientId, permission) =>
  createGDS(sq.from`client_permissions`)({
    clientId,
    userId,
    permission
  });

export const createTestCycle = async (
  clientId,
  cycleName,
  cycleCategory,
  engagementDates
) =>
  createGDS(sq.from`cycles`)({
    clientId,
    name: cycleName || faker.lorem.sentence(),
    category: cycleCategory,
    engagementDates: engagementDates || `[2018-01-01,2018-12-31)`
  });

export const createTestIO = async (
  insertion_order_id,
  client_id,
  created_by,
  attribution,
  campaign_name,
  campaign_description,
  program_phase,
  commission_rate = 0.15
) =>
  createGDS(sq.from`insertion_orders_revisions`)({
    insertion_order_id: insertion_order_id || faker.random.uuid(),
    client_id,
    created_by,
    attribution: attribution || faker.name.firstName(),
    campaign_name: campaign_name || faker.lorem.words(4),
    campaign_description: campaign_description || faker.lorem.word(),
    program_phase: program_phase || faker.lorem.word(),
    commission_rate
  });

export const createTestIOLineItem = async (
  insertion_order_line_item_id,
  insertion_order_id,
  created_by,
  category,
  platform,
  objective,
  cost_gross,
  cost_net,
  active_dates = "[2018-01-01, 2018-12-31)",
  commission_rate = 0.15
) =>
  createGDS(sq.from`insertion_orders_line_items_revisions`)({
    insertion_order_line_item_id:
      insertion_order_line_item_id || faker.random.uuid(),
    insertion_order_id,
    category: category || faker.lorem.word(),
    platform: platform || faker.lorem.word(),
    objective: objective || faker.lorem.word(),
    active_dates,
    cost_gross:
      cost_gross || faker.random.number({ min: 0, max: 100, precision: 2 }),
    cost_net:
      cost_net || faker.random.number({ min: 0, max: 100, precision: 2 }),
    commission_rate,
    created_by
  });
