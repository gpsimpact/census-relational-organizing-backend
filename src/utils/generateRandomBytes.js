import { randomBytes } from "crypto";
import { promisify } from "util";

export default async length => {
  const randomBytesPromiseified = promisify(randomBytes);
  return (await randomBytesPromiseified(length)).toString("hex");
};
