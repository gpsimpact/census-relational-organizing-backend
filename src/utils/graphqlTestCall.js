import { graphql } from "graphql";
import { makeExecutableSchema } from "graphql-tools";
import { applyMiddleware } from "graphql-middleware";
import transformsMiddleware from "../middleware/transforms";
import permissionsMiddleware from "../middleware/permissions";

import makeContext from "../context";
import { default as typeDefs } from "../typeDefinitions";
import resolvers from "../resolvers";

export const schema = makeExecutableSchema({
  typeDefs,
  resolvers
});

const schemaWithMiddleware = applyMiddleware(
  schema,
  transformsMiddleware,
  permissionsMiddleware
);

export const makeTestContext = (userId, contextOverRides) => {
  const context = makeContext(
    {
      session: {
        userId,
        destroy: () => {}
      }
    },
    {
      clearCookie: () => {}
    }
  );
  const testContext = Object.assign(context, contextOverRides);
  return testContext;
};

export const graphqlTestCall = async (
  query,
  variables,
  userId,
  contextOverRides
) => {
  // const schema = await generateSchema();
  const testContext = makeTestContext(userId, contextOverRides);
  return graphql(
    schemaWithMiddleware,
    query,
    undefined,
    testContext,
    variables
  );
};
