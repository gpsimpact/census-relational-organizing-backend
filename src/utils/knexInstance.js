import Knex from "knex";

const knexfile = require("../../knexfile");

export default Knex({
  ...knexfile
});
