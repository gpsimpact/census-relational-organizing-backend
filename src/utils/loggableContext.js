import _ from "lodash";

// this pares down context into something more loggable
export default context => {
  const outContext = _.pick(context, ["user"]);

  outContext.req = _.assign(
    {},
    _.pick(context.req, [
      "headers",
      "url",
      "method",
      "statusCode",
      "statusMessage",
      "baseUrl",
      "originalUrl",
      "params",
      "query",
      "body"
    ])
  );

  return outContext;
};
