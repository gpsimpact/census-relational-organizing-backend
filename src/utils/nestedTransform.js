import flatten from "obj-flatten";
import { keys, endsWith, set, get } from "lodash";

export default (object, targetKey, transformFunction) => {
  const flatObject = flatten(object);
  keys(flatObject).forEach(key => {
    if (endsWith(key, `.${targetKey}`) || endsWith(key, `${targetKey}`)) {
      set(object, key, transformFunction(get(object, key)));
    }
  });
  return object;
};
