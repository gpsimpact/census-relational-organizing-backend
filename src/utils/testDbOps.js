// import { map } from "lodash";
import knex from "./knexInstance";
import redis from "../redis";

// const tableList = [
//   "users",
//   "clients",
//   "global_permissions",
//   "client_permissions",
//   "insertion_orders_current",
//   "insertion_orders_revisions",
//   "insertion_orders_line_items_current",
//   "insertion_orders_line_items_revisions"
// ];

export const dbUp = async () => {
  await knex.raw(`
    TRUNCATE TABLE users CASCADE;
    TRUNCATE TABLE clients CASCADE;
    TRUNCATE TABLE global_permissions CASCADE;
    TRUNCATE TABLE client_permissions CASCADE;
    TRUNCATE TABLE insertion_orders_current CASCADE;
    TRUNCATE TABLE insertion_orders_revisions CASCADE;
    TRUNCATE TABLE insertion_orders_line_items_current CASCADE;
    TRUNCATE TABLE insertion_orders_line_items_revisions CASCADE;
  `);
  await redis.flushdb();
};
// await Promise.all(
//   knex.raw(`
//     TRUNCATE TABLE users CASCADE;
//     TRUNCATE TABLE clients CASCADE;
//     TRUNCATE TABLE global_permissions CASCADE;
//     TRUNCATE TABLE client_permissions CASCADE;
//     TRUNCATE TABLE insertion_orders_current CASCADE;
//     TRUNCATE TABLE insertion_orders_revisions CASCADE;
//     TRUNCATE TABLE insertion_orders_line_items_current CASCADE;
//     TRUNCATE TABLE insertion_orders_line_items_revisions CASCADE;
//   `),
//   redis.flushdb()
// );

export const dbDown = async () => {
  // await Promise.all(
  //   map(tableList, table => knex.raw(`TRUNCATE TABLE ${table} CASCADE;`)),
  //   redis.flushdb()
  // );
};
