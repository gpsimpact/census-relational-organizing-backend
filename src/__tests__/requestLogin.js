import faker from "faker";
import { graphqlTestCall } from "../utils/graphqlTestCall";
import { dbUp } from "../utils/testDbOps";
import { createTestUser } from "../utils/createTestEntities";

const requestLoginMutation = `
    mutation requestLogin($email: String!) {
      requestLogin(email: $email) {
        code
        message
        success
        securityCode
      }
    }
`;

beforeEach(async () => {
  await dbUp();
});

// afterEach(async () => {
//   await dbDown();
// });

describe("RequestLoginResolver", () => {
  test("Non existant user returns DOES_NOT_EXIST and success: false", async () => {
    const response = await graphqlTestCall(requestLoginMutation, {
      email: faker.internet.email()
    });
    expect(response.data.requestLogin.code).toBe("DOES_NOT_EXIST");
    expect(response.data.requestLogin.success).toBe(false);
  });

  test("happy path", async () => {
    const user = await createTestUser({
      email: faker.internet.email().toLowerCase()
    });

    const mockSendEmail = jest.fn();
    const response = await graphqlTestCall(
      requestLoginMutation,
      {
        email: user.email
      },
      {
        sendEmail: mockSendEmail
      }
    );
    expect(mockSendEmail).toHaveBeenCalled();
    expect(response.data.requestLogin.code).toBe("OK");
    expect(response.data.requestLogin.message).toBe(
      "A message has been sent to your email with a magic link you need to click to log in."
    );
    expect(response.data.requestLogin.success).toBe(true);
  });

  test("case trim coercion", async () => {
    const user = await createTestUser({
      email: faker.internet.email().toLowerCase()
    });

    const mockSendEmail = jest.fn();
    const response = await graphqlTestCall(
      requestLoginMutation,
      {
        email: ` ${user.email.toUpperCase()}    `
      },
      {
        sendEmail: mockSendEmail
      }
    );
    expect(mockSendEmail).toHaveBeenCalled();
    expect(response.data.requestLogin.code).toBe("OK");
    expect(response.data.requestLogin.message).toBe(
      "A message has been sent to your email with a magic link you need to click to log in."
    );
    expect(response.data.requestLogin.success).toBe(true);
  });

  test("throws error if logged in", async () => {
    const user = await createTestUser();

    const res = await graphqlTestCall(
      requestLoginMutation,
      {
        email: ` ${user.email.toUpperCase()}    `
      },
      { user: { id: user.id } }
    );
    expect(res.errors.length).toBe(1);
    expect(res.errors[0].message).toEqual("You are already authenticated");
  });
});
