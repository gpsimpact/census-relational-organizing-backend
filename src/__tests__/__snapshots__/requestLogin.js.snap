// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestLoginResolver throws error if logged in 1`] = `
Array [
  [GraphQLError: Not Authorized!],
]
`;
